### ISSUE TITLE
P2587R3 Redefining `to_string` To Use `to_chars`

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#3805]
WG21-P2587R3 Redefining `to_string` To Use `to_chars`

Feature-test macro:
```cpp
#define __cpp_lib_to_string 202306L
```

Note: We're focused on implementing [the remaining library-only features in C++23](https://github.com/microsoft/STL/projects/8?card_filter_query=-label%3Acompiler+-label%3Afixed+-label%3Aresolved). Until that's done, we will **NOT** be accepting PRs for C++26 features.

**Comment by owacoder** [comment]
I'd be interested in working on this.

**Comment by StephanTLavavej** [comment]
Sure, I've moved this to Investigating for you. (As I've mentioned elsewhere, I do plan to open up the repo to C++26 PRs soon, although ideally we should land `feature/flat_map` first.)

**Comment by owacoder** [comment]
Since using std::format as described by the paper is really unnecessary, and only a dependency on `to_chars` is needed, is it possible to just include `<charconv>` instead of `<format>`? Otherwise, there will be a circular dependency where format needs string and string needs format.

As the original author of [this answer](https://stackoverflow.com/questions/68472720/stdto-chars-minimal-floating-point-buffer-size/68475665#68475665) I would also like to include/relicense/contribute my code into the implementation to allocate appropriately ahead of time (documenting here so there's no licensing issues).

For wide chars, either a conversion will be needed or a dependency on format. I realize the conversion can be a simple widening of each character since the set of characters is limited and well-defined, but don't know which approach is preferable. Would you be able to advise?

**Comment by frederick-vs-ja** [comment]
> Since using `format` as described by the paper is really unnecessary, and only a dependency on `to_chars` is needed, is it possible to just include `<charconv>` instead of `<format>`?

Yes.

> Otherwise, there will be a circular dependency where format needs string and string needs format.

Even if we use `std::format` in `std::to_string` as indicated by the standard wording, the circular dependency can be avoided by splitting headers, because we don't use `std::to_string` in `<format>`.

> For wide chars, either a conversion will be needed or a dependency on format. I realize the conversion can be a simple widening of each character since the set of characters is limited and well-defined, but don't know which approach is preferable. Would you be able to advise?

I think the latter is preferable. Actually, in `<format>` we're already performing such simple widening.

**Comment by owacoder** [comment]
It seems a severe pessimization to call out to `std::format("{}", _Val)` instead of just calling `std::to_chars` directly. The parsing overhead for the format string would be completely unnecessary. Is it permissible to call `std::to_chars` under an "as-if" rule since it has the same behavior?

Given the title of the revision paper (and the simplicity of the underlying implementation) it would appear to me that's actually acceptable.

**Comment by StephanTLavavej** [comment]
Yes, I believe it is permissible to call `to_chars` via "as if". It's because there's no way for users to affect how `format` calls `to_chars` here.

**Comment by owacoder** [comment]
This is now implemented here: [https://github.com/owacoder/STL/tree/p2587r3](https://github.com/owacoder/STL/tree/p2587r3).

However, the new implementation requires that `<string>` pull in `<charconv>`. I don't know the STL procedures well enough to know if that is okay or if more code needs to be extracted out.

I also have not added test coverage yet.

**Comment by frederick-vs-ja** [comment]
I think we can include `<charconv>` in `<string>` only since C++26. Perhaps we can split the `to_chars` part improve throughput. Also, it seem that some stuffs should be inlined as it's used only once.

**Comment by AlexDeFoc** [comment]
any updates? the repo doesn't show anymore commits since a year ago

**Comment by owacoder** [comment]
Just waiting for C++26 floodgates to open. It just needs test coverage, otherwise it should be ready to go.
